[WASM-Function-References] Add support for local initialization tracking

takikawa opened the pull request:
#### 83c1c1073f0f9d01b9bb573d3cf50aee14314d5a
<pre>
[WASM-Function-References] Add support for local initialization tracking
<a href="https://bugs.webkit.org/show_bug.cgi?id=247396">https://bugs.webkit.org/show_bug.cgi?id=247396</a>

Reviewed by Justin Michaud.

Add local initialization tracking as added by the typed function references
proposal. This allows the use of non-defaultable locals (e.g., with types like
(ref extern) that are non-nullable). Non-defaultable locals can only be read
with local.get after an initialization by, e.g., local.set.

To track the local state, an initialization stack (max height is the number of
non-defaultable locals) is used that tracks initialized non-defaultable locals.
The stack is popped to a previous state on exit from a block.  A bitvector is
used to map local indices to initialization state, to facilitate quick lookups.

* JSTests/wasm/function-references-spec-tests/local_init.wast.js: Added.
* JSTests/wasm/function-references/local_init.js: Added.
(module):
(async testLocalInit):
* JSTests/wasm/function-references/ref_types.js:
(async testNonNullRefTypeLocal):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser::pushLocalInitialized):
(JSC::Wasm::FunctionParser::getLocalInitStackHeight const):
(JSC::Wasm::FunctionParser::resetLocalInitStackToHeight):
(JSC::Wasm::FunctionParser::localIsInitialized):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parse):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseBody):
(JSC::Wasm::FunctionParser&lt;Context&gt;::checkLocalInitialized):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseUnreachableExpression):

Canonical link: <a href="https://commits.webkit.org/267309@main">https://commits.webkit.org/267309@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ae2d3b1e49af2e3aeb9e0cf73b5d9118e126b53a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16197 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16514 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16926 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17962 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15201 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19574 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16625 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17593 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16391 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16842 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13841 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18728 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14085 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14659 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21499 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/13949 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15074 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14824 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/18059 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/15435 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15419 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13072 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/16399 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14644 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/4086 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3882 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19010 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/17565 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15241 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3878 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -137,6 +138,23 @@ class FunctionParser : public Parser<void>, public FunctionParserTypes<typename
+    void pushLocalInitialized(uint32_t index)
+    {
+        if (Options::useWebAssemblyTypedFunctionReferences() && !isDefaultableType(typeOfLocal(index)) && !localIsInitialized(index)) {
+            m_localInitStack.uncheckedAppend(index);
```
Why is this unchecked?

takikawa replied:
In the `parse()` function: https://github.com/WebKit/WebKit/blob/56633a9f762503fd20ea2be164489d70b9cf4083/Source/JavaScriptCore/wasm/WasmFunctionParser.h#L362 the `m_localInitStack` is sized to be the number of non-defaultable locals. Since we only ever push non-defaultable locals to this stack (and we only push if it's not already on the stack), it's not supposed to exceed that size.

justinmichaud replied:
Let's just leave it as checked. Unchecked is only supposed to be for super perf critical things, or things where right above you can see why it doesn't need a check

takikawa replied:
Ok no problem, will change it to normal checked `append(index)`.

justinmichaud reviewed (approved):
r=me

takikawa committed:
[WASM-Function-References] Add support for local initialization tracking
https://bugs.webkit.org/show_bug.cgi?id=247396

Reviewed by Justin Michaud.

Add local initialization tracking as added by the typed function references
proposal. This allows the use of non-defaultable locals (e.g., with types like
(ref extern) that are non-nullable). Non-defaultable locals can only be read
with local.get after an initialization by, e.g., local.set.

To track the local state, an initialization stack (max height is the number of
non-defaultable locals) is used that tracks initialized non-defaultable locals.
The stack is popped to a previous state on exit from a block.  A bitvector is
used to map local indices to initialization state, to facilitate quick lookups.

* JSTests/wasm/function-references-spec-tests/local_init.wast.js: Added.
* JSTests/wasm/function-references/local_init.js: Added.
(module):
(async testLocalInit):
* JSTests/wasm/function-references/ref_types.js:
(async testNonNullRefTypeLocal):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser::pushLocalInitialized):
(JSC::Wasm::FunctionParser::getLocalInitStackHeight const):
(JSC::Wasm::FunctionParser::resetLocalInitStackToHeight):
(JSC::Wasm::FunctionParser::localIsInitialized):
(JSC::Wasm::FunctionParser<Context>::parse):
(JSC::Wasm::FunctionParser<Context>::parseBody):
(JSC::Wasm::FunctionParser<Context>::checkLocalInitialized):
(JSC::Wasm::FunctionParser<Context>::parseExpression):
(JSC::Wasm::FunctionParser<Context>::parseUnreachableExpression):

Canonical link: https://commits.webkit.org/267309@main
